### DISCUSSION TITLE
Soundness concerns

### DISCUSSION THREAD

**Discussion by seanpianka** [routerify/routerify#46] [General]
Reposting my concerns from #45:

How should the caller guarantee the lifetimes? The RequestService can reach a Router, but the lifetime is not tied to that of the Router. This could easily cause a use-after-free or stack-use-after-return.

Similarly, does unsafe impl Send + unsafe impl Sync add data races? I do not understand how this is sound, since the Router does not have to be thread-safe.

Also, I asked a while ago about wrapping the `Router` in an `Arc`, and @rousan mentioned there are performances issues:

> Actually, because of the performance. If I have used Arc, it will impact the performance very badly when you will have a lot of simultaneous requests, as the Arc manages a concurrent counter to keep a track of the current active owners and this counter gets updated only once at a time, so if you got 1k requests, this counter will get updated one by one and other new requests will wait which impacts the performance.

The confusion here is that the reference count is incremented when you deference the Arc. This is not true -- you only increment the reference count when the Arc is cloned. When you deference the Arc, you're only checking a (thread-)local, or don't check it all (I forgot the specifics).

Unless there was a test done which showed a degradation of performance using `Arc` vs `mut*`, I'd like to discuss making this change.

**Comment by gsserge** [comment]
I'm still mastering Rust, and the lifetimes + sharing part is one of the toughest so far. That said, it's my understanding that there's absolutely no way of sharing a router instance by reference between threads without either wrapping it in Arc or using unsafe. 

You are right that technically router does not have to be thread-safe. Currently, however, the router is passed by a **mutable** raw pointer because of the `FnMut` trait bound on the handlers and middleware. I've done a little experiment converting `FnMut` to `Fn` - that allowed me to switch from a mutable raw pointer to an immutable one. So now the router object is definitely thread safe. 

The problem though is not the (im)mutability and thread safety, but **liveliness**. There seem to be no way to come up with the right combination of filetimes for `RouterService` and `RequestService` to convince the compiler that the router instance owned by `RouterService` and passed by reference to `RequestService` will live long enough to be safely processed by Tokio tasks. Tokio will only allow tasks to execute functions or closures with `'static` trait bound, meaning all data inside a fn/closure must be owned by it. This kind of makes sense, but I wish there was a way to hint the compiler that `Server::serve()` will run as long as the whole app is running and whatever is allocated before the `serve()` call, and immutably passed into it by ref, will not get dropped.

It seems like the only way to deal with this safely is to wrap the router in Arc. Which is definitely going to hurt performance, but does it really matter? It might for simple synthetic benchmarks, but any real-world web API is very likely going to be bound by the latency of the database or queue or some third-party services.

@seanpianka I agree that it would be better and safer to resurrect https://github.com/routerify/routerify/pull/25 despite the performance drop. @rousan what do you think?

**Reply by seanpianka** [reply]
#50 resolves this issue, thank you for your help!
